packages/compiler/core/semantic-walker.ts

```diff
@@ -63,171 +108,121 @@ function checkVisited(visited: Set<any>, item: any) {
     return;
   }
 
-  eventEmitter.emit("interface", type);
+  context.emitter.emit("interface", type);
   for (const op of type.operations.values()) {
-    navigateType(op, eventEmitter, visited);
+    navigateOperationType(op, context);
   }
 }
 
-function navigateEnumType(
-  type: Enum,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
-  if (checkVisited(visited, type)) {
+function navigateEnumType(type: Enum, context: NavigationContext) {
+  if (checkVisited(context.visited, type)) {
     return;
   }
 
-  eventEmitter.emit("enum", type);
+  context.emitter.emit("enum", type);
 }
 
-function navigateUnionType(
-  type: Union,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
-  if (checkVisited(visited, type)) {
+function navigateUnionType(type: Union, context: NavigationContext) {
+  if (checkVisited(context.visited, type)) {
+    return;
+  }
+  if (!context.options.includeTemplateDeclaration && isTemplateDeclaration(type)) {
     return;
   }
-  eventEmitter.emit("union", type);
+  context.emitter.emit("union", type);
   for (const variant of type.variants.values()) {
-    navigateType(variant, eventEmitter, visited);
+    navigateReferencedType(variant, context);
   }
 }
 
-function navigateUnionTypeVariant(
-  type: UnionVariant,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
-  if (checkVisited(visited, type)) {
+function navigateUnionTypeVariant(type: UnionVariant, context: NavigationContext) {
+  if (checkVisited(context.visited, type)) {
     return;
   }
-  eventEmitter.emit("unionVariant", type);
-  navigateType(type.type, eventEmitter, visited);
+  context.emitter.emit("unionVariant", type);
+  navigateReferencedType(type.type, context);
 }
 
-function navigateTupleType(
-  type: Tuple,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
-  if (checkVisited(visited, type)) {
+function navigateTupleType(type: Tuple, context: NavigationContext) {
+  if (checkVisited(context.visited, type)) {
     return;
   }
-  eventEmitter.emit("tuple", type);
+  context.emitter.emit("tuple", type);
   for (const value of type.values) {
-    navigateType(value, eventEmitter, visited);
+    navigateReferencedType(value, context);
   }
 }
 
-function navigateTemplateParameter(
-  type: TemplateParameter,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
-  if (checkVisited(visited, type)) {
-    return;
-  }
-  eventEmitter.emit("templateParameter", type);
-}
+// function navigateTemplateParameter(type: TemplateParameter, context: NavigationContext) {
+//   if (checkVisited(context.visited, type)) {
+//     return;
+//   }
+//   context.emitter.emit("templateParameter", type);
+// }
 
-function navigateType(
-  type: Type,
-  eventEmitter: EventEmitter<SemanticNodeListener>,
-  visited: Set<any>
-) {
+function navigateReferencedType(type: Type, context: NavigationContext) {
```

Comment:
Updated here to call this when dealing with a referenced type(like operation return type, model property type, etc.)

The only time we should keep navigate for those type is when they were created via expression(model expression and union expression). Otherwise those should be availaible via another mean.

Reply:
@bterlson do you see any issues with this approach.